README.md
# Spitfire

![Spitfire: Seamlessly create videos from images](https://raw.githubusercontent.com/seanmcneil/Spitfire/master/spitfire.jpg)

[![CI Status](http://img.shields.io/travis/seanmcneil/Spitfire.svg?style=flat)](https://travis-ci.org/seanmcneil/Spitfire)
[![Version](https://img.shields.io/cocoapods/v/Spitfire.svg?style=flat)](http://cocoapods.org/pods/Spitfire)
[![License](https://img.shields.io/cocoapods/l/Spitfire.svg?style=flat)](http://cocoapods.org/pods/Spitfire)
[![Platform](https://img.shields.io/cocoapods/p/Spitfire.svg?style=flat)](http://cocoapods.org/pods/Spitfire)
[![CocoaPods](https://img.shields.io/cocoapods/dt/Spitfire.svg)](http://cocoapods.org/pods/Spitfire)

Spitfire is a simple utility for taking an array of images and creating a video from them.

## Current Status

All previous beta versions have been removed from the repository, as they do not comply with the current design.

## Example

To run the example project, clone the repo, and run `pod install` from the Example directory first.

The following code can be found in the example project, but demonstrates how to call it and handle the various callbacks it supports:

```swift
let spitfire = Spitfire()
```

And a function for creating the video:

```swift
func createVideo(for images: [UIImage]) {
    do {
        try spitfire.makeVideo(with: images, progress: { (progress) in
            let percent = (progress.fractionCompleted * 100).roundTo(places: 2)
            print("\(percent)%")
        }, success: { (url) in
            PHPhotoLibrary.shared().performChanges({
            PHAssetChangeRequest.creationRequestForAssetFromVideo(atFileURL: url)
            }) { saved, error in
                if saved {
                    let alertController = UIAlertController(title: NSLocalizedString("Your video was saved", comment: ""), message: nil, preferredStyle: .alert)
                    let defaultAction = UIAlertAction(title: NSLocalizedString("OK", comment: ""), style: .default, handler: nil)
                    alertController.addAction(defaultAction)
                    self.present(alertController, animated: true, completion: nil)
                }
            }
        })
    } catch {
        print(error.localizedDescription)
    }
}
```

## Performance Considerations

Be aware that the image array to feed the writer can start to get immense, easily exceeding 1GB of RAM

## Requirements
- iOS 11.0+
- Xcode 9.0+
- Swift 4.0+

## Installation

Spitfire is available through [CocoaPods](http://cocoapods.org). To install
it, simply add the following line to your Podfile:

```ruby
pod "Spitfire"
```

## Usage

```swift
import Spitfire
```

Ensure that you declare the Spitfire property at the class level so that it does not go out of scope during execution:

```swift
class MyClass {
    let spitfire = Spitfire()
    ...
```

Call the default makeVideo function, which uses a value of 30 for frame rate:

```swift
do {
    try spitfire.makeVideo(with: images, progress: { (progress) in
        // Update user on status
    }, success: { (url) in
        // Process the completed video
    })
} catch {
    // Handle any errors
}
``` 

Call the optional makeVideo function and specify a frame rate between 1 and 60:

```swift
do {
    try spitfire.makeVideo(with: images, fps: 5, progress: { (progress) in
        // Update user on status
    }, success: { (url) in
        // Process the completed video
    })
} catch {
    // Handle any errors
}
``` 

## Errors

Spitfire provides a relatively rich set of errors via an enum that should address all potential failures within the app. These are:

```swift
public enum SpitfireError: Swift.Error {
    case ImageArrayEmpty
    case InvalidFramerate(String)
    case ImageDimensionsMatchFailure
    case ImageDimensionsMultiplierFailure(String)
    case VideoWriterFailure
    case PixelBufferPointeeFailure
    case InvalidStatusCode(Int)
    case PixelBufferApendFailure
}
```

## Credits

This work is based off of work originally performed by [acj](https://gist.github.com/acj) which can be found [here](https://gist.github.com/acj/6ae90aa1ebb8cad6b47b).

## License

Spitfire is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
## 1.1.0 Release (2019-03-25)

##### Enhancements

## 1.1.0 Release
* Compatibility with Swift 4.0

## 1.0.1 Release
* Refactor code related to PixelBuffer class

* Added some additional comments

* Improved variable and function naming


## 1.0.0 Release
* Initial release

* Beta versions retired

* Design now relies on a try / catch block to handle errors versus a failure block

* Improved readme


Example/Tests/Tests.swift
import UIKit
import XCTest
import Spitfire

class Tests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testExample() {
        // This is an example of a functional test case.
        XCTAssert(true, "Pass")
    }
    
    func testPerformanceExample() {
        // This is an example of a performance test case.
        self.measure() {
            // Put the code you want to measure the time of here.
        }
    }
    
}


Example/Tests/SpitfireTests.swift
//
//  SpitfireTests.swift
//  Spitfire
//
//  Created by seanmcneil on 3/26/17.
//  Copyright © 2017 CocoaPods. All rights reserved.
//

import XCTest
import UIKit

@testable import Spitfire

class SpitfireTests: XCTestCase {
    func testInitEmptyArray() {
        let spitfire = Spitfire()
        let emptyArray = [UIImage]()
        
        XCTAssertThrowsError(try spitfire.makeVideo(with: emptyArray, progress: { (progress) in
            XCTFail()
        }, success: { (url) in
            XCTFail()
        })) { error -> Void in
            switch error {
            case SpitfireError.ImageArrayEmpty:
                break
            default:
                XCTFail("Incorrect error type")
            }
        }
    }
    
    func testFrameRateTooLow() {
        let spitfire = Spitfire()
        let lowFPS: Int32 = 0
        let array = [UIImage()]
        
        XCTAssertThrowsError(try spitfire.makeVideo(with: array, fps:lowFPS, progress: { (progress) in
            XCTFail()
        }, success: { (url) in
            XCTFail()
        })) { error -> Void in
            switch error {
            case SpitfireError.InvalidFramerate("Framerate must be between 1 and 60"):
                break
            default:
                XCTFail("Incorrect error type")
            }
        }
    }
    
    func testFrameRateTooHigh() {
        let spitfire = Spitfire()
        let lowFPS: Int32 = 61
        let array = [UIImage()]
        
        XCTAssertThrowsError(try spitfire.makeVideo(with: array, fps:lowFPS, progress: { (progress) in
            XCTFail()
        }, success: { (url) in
            XCTFail()
        })) { error -> Void in
            switch error {
            case SpitfireError.InvalidFramerate("Framerate must be between 1 and 60"):
                break
            default:
                XCTFail("Incorrect error type")
            }
        }
    }
    
    func testDivisbleBy16() {
        let spitfire = Spitfire()
        let images: [UIImage] = [UIImage(color: .black)!]
        
        XCTAssertThrowsError(try spitfire.makeVideo(with: images, progress: { (progress) in
            XCTFail()
        }, success: { (url) in
            XCTFail()
        })) { error -> Void in
            switch error {
            case SpitfireError.ImageDimensionsMultiplierFailure("Image width must be divisble by 16"):
                break
            default:
                XCTFail("Incorrect error type")
            }
        }
    }


    
    func testPerformanceExample() {
        // This is an example of a performance test case.
        self.measure {
            // Put the code you want to measure the time of here.
        }
    }
    
}

public extension UIImage {
    public convenience init?(color: UIColor, size: CGSize = CGSize(width: 1, height: 1)) {
        let rect = CGRect(origin: .zero, size: size)
        UIGraphicsBeginImageContextWithOptions(rect.size, false, 0.0)
        color.setFill()
        UIRectFill(rect)
        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        
        guard let cgImage = image?.cgImage else { return nil }
        
        self.init(cgImage: cgImage)
    }
}


Example/Spitfire/AppDelegate.swift
//
//  AppDelegate.swift
//  Spitfire
//
//  Created by seanmcneil on 03/08/2017.
//  Copyright (c) 2017 seanmcneil. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?


    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        return true
    }

    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(_ application: UIApplication) {
        // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }


}



Spitfire/Classes/PixelBuffer.swift
//
//  PixelBuffer.swift
//  Pods
//
//  Created by seanmcneil on 3/26/17.
//
//

import AVFoundation

extension Spitfire {
    // Set up pixel buffer to add a frame at specified time
    func append(pixelBufferAdaptor adaptor: AVAssetWriterInputPixelBufferAdaptor, with image: UIImage, at presentationTime: CMTime, success: @escaping (() -> ())) throws {
        do {
            if let pixelBufferPool = adaptor.pixelBufferPool {
                let pixelBufferPointer = UnsafeMutablePointer<CVPixelBuffer?>.allocate(capacity: 1)
                let status = CVPixelBufferPoolCreatePixelBuffer(kCFAllocatorDefault, pixelBufferPool, pixelBufferPointer)
                guard let pixelBuffer = pixelBufferPointer.pointee else {
                    throw(SpitfireError.PixelBufferPointeeFailure)
                }
                guard status == 0 else {
                    throw(SpitfireError.InvalidStatusCode(Int(status)))
                }
                
                fill(pixelBuffer: pixelBuffer, with: image)
                if adaptor.append(pixelBuffer, withPresentationTime: presentationTime) {
                    pixelBufferPointer.deinitialize(count: 1)
                    pixelBufferPointer.deallocate()
                    success()
                } else {
                    throw(SpitfireError.PixelBufferApendFailure)
                }
            }
        } catch let error {
            throw error
        }
    }
    
    // Populates the pixel buffer with the contents of the current image
    private func fill(pixelBuffer buffer: CVPixelBuffer, with image: UIImage) {
        CVPixelBufferLockBaseAddress(buffer, CVPixelBufferLockFlags(rawValue: CVOptionFlags(0)))
        
        guard let context = CGContext(
            data: CVPixelBufferGetBaseAddress(buffer),
            width: Int(image.size.width),
            height: Int(image.size.height),
            bitsPerComponent: 8,
            bytesPerRow: CVPixelBufferGetBytesPerRow(buffer),
            space: CGColorSpaceCreateDeviceRGB(),
            bitmapInfo: CGImageAlphaInfo.premultipliedFirst.rawValue
            ) else { return }
        guard let cgImage = image.cgImage else { return }
        
        let rect = CGRect(origin: .zero, size: image.size)
        context.draw(cgImage, in: rect)
        
        CVPixelBufferUnlockBaseAddress(buffer, CVPixelBufferLockFlags(rawValue: CVOptionFlags(0)))
    }
}


Example/Spitfire/ViewController.swift
//
//  ViewController.swift
//  Spitfire
//
//  Created by seanmcneil on 03/08/2017.
//  Copyright (c) 2017 seanmcneil. All rights reserved.
//

import UIKit
import Spitfire
import Photos

class ViewController: UIViewController {
    let spitfire = Spitfire()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        createVideo()
    }

    func createVideo() {
        var images = [UIImage]()
        for _ in 0..<120 {
            let image = UIImage(named: "testImage")!
            images.append(image)
        }
        
        do {
            try spitfire.makeVideo(with: images, progress: { (progress) in
                let percent = (progress.fractionCompleted * 100).roundTo(places: 2)
                print("\(percent)%")
            }, success: { (url) in
                PHPhotoLibrary.shared().performChanges({
                    PHAssetChangeRequest.creationRequestForAssetFromVideo(atFileURL: url)
                }) { saved, error in
                    if saved {
                        let alertController = UIAlertController(title: NSLocalizedString("Your video was saved", comment: ""), message: nil, preferredStyle: .alert)
                        let defaultAction = UIAlertAction(title: NSLocalizedString("OK", comment: ""), style: .default, handler: nil)
                        alertController.addAction(defaultAction)
                        self.present(alertController, animated: true, completion: nil)
                    }
                }
            })
        } catch {
            print(error.localizedDescription)
        }
    }
}

extension Double {
    func roundTo(places:Int) -> Double {
        let divisor = pow(10.0, Double(places))
        return (self * divisor).rounded() / divisor
    }
}


Spitfire/Classes/Spitfire.swift
//
//  Spitfire.swift
//  Pods
//
//  Created by seanmcneil on 3/8/17.
//
//

import AVFoundation
import UIKit

public class Spitfire {
    public init() { }
    
    private var videoWriter: AVAssetWriter?
    
    private var outputURL: URL {
        let documentsPath = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)[0]
        let documentURL = URL(fileURLWithPath: documentsPath)
        
        return documentURL.appendingPathComponent("output.mov")
    }
    
    /// Produces a video based on the contents of a UIImage array
    ///
    /// - Parameters:
    ///   - images: Images to use for creating video. Should all have the same dimensions
    ///   - fps: Frames per second, with a default value of 30
    ///   - progress: Handler that will return a fractional value indicating percent complete
    ///   - success: Handler that will return a URL of the completed video if successful
    ///   - failure: Handler that will return an error message if one occurs
    public func makeVideo(with images: [UIImage], fps: Int32 = 30, progress: @escaping ((Progress) -> ()), success: @escaping ((URL) -> ())) throws {
        guard let size = images.first?.size else {
            throw(SpitfireError.ImageArrayEmpty)
        }
        
        guard fps > 0 && fps <= 60 else {
            let message = NSLocalizedString("Framerate must be between 1 and 60", comment: "")
            throw(SpitfireError.InvalidFramerate(message))
        }
        
        guard (size.width .truncatingRemainder(dividingBy: 16.0)) == 0 else {
            let message = NSLocalizedString("Image width must be divisble by 16", comment: "")
            throw(SpitfireError.ImageDimensionsMultiplierFailure(message))
        }
        
        try? FileManager.default.removeItem(at: outputURL)
        
        do {
            try videoWriter = AVAssetWriter(outputURL: outputURL, fileType: AVFileType.mov)
        } catch let error {
            throw(error)
        }
        
        guard let videoWriter = videoWriter else {
            throw(SpitfireError.VideoWriterFailure)
        }
        
        let videoSettings: [String : Any] = [
            AVVideoCodecKey  : AVVideoCodecType.h264,
            AVVideoWidthKey  : size.width,
            AVVideoHeightKey : size.height,
        ]
        
        let videoWriterInput = AVAssetWriterInput(mediaType: AVMediaType.video, outputSettings: videoSettings)
        
        let sourceBufferAttributes: [String : Any] = [
            (kCVPixelBufferPixelFormatTypeKey as String): Int(kCVPixelFormatType_32ARGB),
            (kCVPixelBufferWidthKey as String): Float(size.width),
            (kCVPixelBufferHeightKey as String): Float(size.height)]
        
        let pixelBufferAdaptor = AVAssetWriterInputPixelBufferAdaptor(
            assetWriterInput: videoWriterInput,
            sourcePixelBufferAttributes: sourceBufferAttributes
        )
        
        assert(videoWriter.canAdd(videoWriterInput))
        videoWriter.add(videoWriterInput)
        
        if videoWriter.startWriting() {
            videoWriter.startSession(atSourceTime: CMTime.zero)
            assert(pixelBufferAdaptor.pixelBufferPool != nil)
            let writeQueue = DispatchQueue(label: "writeQueue")
            videoWriterInput.requestMediaDataWhenReady(on: writeQueue, using: { [weak self] in
                let frameDuration = CMTimeMake(value: 1, timescale: fps)
                let currentProgress = Progress(totalUnitCount: Int64(images.count))
                var frameCount: Int64 = 0
                
                while(Int(frameCount) < images.count) {
                    // Will continue to loop until the video writer is able to write, which effectively handles buffer backups
                    if videoWriterInput.isReadyForMoreMediaData {
                        let lastFrameTime = CMTimeMake(value: frameCount, timescale: fps)
                        let presentationTime = frameCount == 0 ? lastFrameTime : CMTimeAdd(lastFrameTime, frameDuration)
                        let image = images[Int(frameCount)]
                        
                        do {
                            guard image.size == size else {
                                throw(SpitfireError.ImageDimensionsMatchFailure)
                            }
                        } catch { } // Do not throw here
                        
                        do {
                            try self?.append(pixelBufferAdaptor: pixelBufferAdaptor, with: image, at: presentationTime, success: {
                                frameCount += 1
                                currentProgress.completedUnitCount = frameCount
                                progress(currentProgress)
                            })
                        } catch { } // Do not throw here
                    }
                }
                
                videoWriterInput.markAsFinished()
                videoWriter.finishWriting { [weak self] () -> Void in
                    guard let strongSelf = self else { return }
                    
                    success(strongSelf.outputURL)
                }
            })
        }
    }
}
